extensions: add http bandwidth limit filter

nitgoy opened the pull request:
Commit Message: adds bandwidth limit filter core functionality and UTs
Additional Description: the filter smoothens the flow of data in both direction up to the specified bandwidth limit.
Risk Level:
Testing: UTs added. Adding more UTs and integration tests.
Docs Changes: Pending
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #13604]
[Optional Deprecated:]

tonya11en commented on api/envoy/extensions/filters/http/bandwidth_limit/v3/bandwidth_limit.proto:
```diff
@@ -0,0 +1,63 @@
+  uint64 limit_kbps = 3 [(validate.rules).uint64 = {gte: 1}];
+
+  // Optional fill_rate for the token buckets (per second). Defaults to 64.
+  google.protobuf.UInt64Value fill_rate = 4 [(validate.rules).uint64 = {gte: 1}];
```
May make more sense to have this be a duration or a time window in ms.

nitgoy replied:
The reason I don't have it as duration type here is because semantically it represents parts of a second of time and not just any time duration. That derives from the fact that the actual limit is implicitly defined in "per second". If I change it to duration, I'll have to convert it to number of sub-parts of a second and ensure config value is never more than a second. For this reason, I chose to not use the duration type. Thoughts?

tonya11en commented on api/envoy/extensions/filters/http/bandwidth_limit/v3/bandwidth_limit.proto:
```diff
@@ -0,0 +1,63 @@
+
+  // If set, the bandwidth limit will be enforced only when total dataplane bandwidth usage
+  // crosses this value. Defaults to 0 meaning limits are always applied.
+  google.protobuf.UInt64Value enforce_threshold_kbps = 5 [(validate.rules).uint64 = {gte: 1}];
```
Is this a way to do a "fair share" of the network when you have excess resources? If so, there are better ways we can achieve the multitanancy (I deal with this exact problem in my day job). Let's omit it from the first PR and chat more about this.

nitgoy replied:
Yes, it's for fair share. I'd love to hear about better ways to achieve it. Let me know.

nitgoy replied:
Did you mean to use the resource monitor/overload manager to detect whether the bandwidth is stressed? If so, that'd be great but I'm not sure how to consume that trigger inside of a filter. I have removed this config for now though.

tonya11en replied:
This is one of those behaviors that should be delegated to its own QoS filter. There are plenty of ways to do this, but it will require some buffering/queuing and it seems out of scope for a bandwidth limiter filter.

tonya11en reviewed (commented):
Looks great! 

You'll want to see how other filters implement their config as an alpha version so the proto isn't locked in.

Let's use this PR for the whole filter review! I'll stand by for the rest of it. Thanks for doing this.

nitgoy commented on api/envoy/extensions/filters/http/bandwidth_limit/v3/bandwidth_limit.proto:
```diff
@@ -0,0 +1,63 @@
+message BandwidthLimit {
+  // Defines the mode for the bandwidth limit filter.
+  // Values represent bitmasks.
+  enum EnableMode {
```
Yes.

nitgoy commented on api/envoy/extensions/filters/http/bandwidth_limit/v3/bandwidth_limit.proto:
```diff
@@ -0,0 +1,63 @@
+  // .. note::
+  //   When using per route configuration, the limit becomes unique to that route.
+  //
+  uint64 limit_kbps = 3 [(validate.rules).uint64 = {gte: 1}];
```
You can define the Gbps/Mbps limits in Kbps. Semantically, it's a whole number so FractionalPercent didn't make sense to me. It's same as the [fault filter config](https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/v3/token_bucket.proto#envoy-v3-api-msg-type-v3-tokenbucket).

nitgoy commented:
Moved the config to alpha version and added generated configs and responded to other comments.

Have a couple of questions:

1. How to enable generation of the v4alpha config? As per this [documentation](https://github.com/envoyproxy/envoy/blob/master/api/STYLE.md#adding-an-extension-configuration-to-the-api) it should be auto generated but somehow it didn't. Am I missing some import?
2. I see the message that says I'm not authorized to merge the PR. How do I obtain permissions to be able to merge PRs into envoy master?
3. This is my first envoy PR and I don't have a bunch of OSS experience. Let me know if there is anything else I should be aware of when contributing to envoy. I already plan to follow the [envoy contribution guide](https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md).

Thanks!

tonya11en commented on api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto:
```diff
@@ -0,0 +1,60 @@
+  // Values represent bitmask.
+  enum EnableMode {
+    // Filter is disabled.
+    Disabled = 0;
```
This should be pulled out into its own field using a RuntimeFeatureFlag. Here's an example of one:
https://github.com/envoyproxy/envoy/blob/853ecaa2f1e7364c1db49dad6a91bfcb3bcc4d61/api/envoy/config/listener/v3/quic_config.proto#L40

You can poke around and see how it would be used in the actual filter. This way folks can toggle it on and off for testing/rollouts/killswitching.

nitgoy replied:
sure, let me look into it.

tonya11en reviewed (changes requested):
> How to enable generation of the v4alpha config? As per this documentation it should be auto generated but somehow it didn't. Am I missing some import?

I believe you need to make **a separate** proto file in `api/envoy/extensions/filters/http/bandwidth_limit/v4alpha/...` to generate those.

> I see the message that says I'm not authorized to merge the PR. How do I obtain permissions to be able to merge PRs into envoy master?

Only senior maintainers can merge Envoy PRs to master. The way it typically works is you go through the whole review process and then a maintainer swoops in to review/merge when it's ready. There's probably more specifics on how it works [here](https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md).

> This is my first envoy PR and I don't have a bunch of OSS experience. Let me know if there is anything else I should be aware of when contributing to envoy. I already plan to follow the envoy contribution guide.

Exciting! Well thank you for making things happen and submitting a PR!

With respect to this actual PR, I think it would make more sense as a network filter instead of an HTTP filter. See [the echo filter](https://github.com/envoyproxy/envoy-filter-example) for a super simple example of one. I'm sure there are several full filter implementation commits in the git history to reference as well. This way traffic can be throttled before HTTP/gRPC messages are parsed.

The proto fields themselves look pretty good to me, modulo the minor comments, so you can feel free to push an actual implementation to the PR and we can start iterating on that.

nitgoy commented:
> > How to enable generation of the v4alpha config? As per this documentation it should be auto generated but somehow it didn't. Am I missing some import?
> 
> I believe you need to make **a separate** proto file in `api/envoy/extensions/filters/http/bandwidth_limit/v4alpha/...` to generate those.
>
I did create a separate proto under v4alpha, following this [proto](https://github.com/envoyproxy/envoy/blob/master/api/envoy/extensions/filters/http/cache/v4alpha/cache.proto). But when I run `proto_format.sh fix`, it deletes those files. When I mark the package_version_status as ACTIVE instead of NEXT_MAJOR_VERSION_CANDIDATE, it works but I'm not sure if that's the right way because it adds both v3 and v4 entries to [api/versioning/BUILD](https://github.com/envoyproxy/envoy/blob/master/api/versioning/BUILD).
 
> > I see the message that says I'm not authorized to merge the PR. How do I obtain permissions to be able to merge PRs into envoy master?
> 
> Only senior maintainers can merge Envoy PRs to master. The way it typically works is you go through the whole review process and then a maintainer swoops in to review/merge when it's ready. There's probably more specifics on how it works [here](https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md).
> 
Got it. I must have missed that part.

> > This is my first envoy PR and I don't have a bunch of OSS experience. Let me know if there is anything else I should be aware of when contributing to envoy. I already plan to follow the envoy contribution guide.
> 
> Exciting! Well thank you for making things happen and submitting a PR!
> 
> With respect to this actual PR, I think it would make more sense as a network filter instead of an HTTP filter. See [the echo filter](https://github.com/envoyproxy/envoy-filter-example) for a super simple example of one. I'm sure there are several full filter implementation commits in the git history to reference as well. This way traffic can be throttled before HTTP/gRPC messages are parsed.
>
I've already implemented it as an http filter, just working on the tests. It was based on the existing http fault filter code, so that helped a lot. I'll push out that code fairly soon.
Although, I'm not sure I fully understand the benefit you stated. Since it's a bandwidth smoothening filter, the messages are supposed to eventually be passed. The filter will only delay that to ensure a smooth bandwidth usage. So whether you do it at L4 or L7 shouldn't really matter, right?
> 
> The proto fields themselves look pretty good to me, modulo the minor comments, so you can feel free to push an actual implementation to the PR and we can start iterating on that.

Thanks for the help. I'll push the implementation soon pending updates to tests and docs.

tonya11en commented:
> Although, I'm not sure I fully understand the benefit you stated. Since it's a bandwidth smoothening filter, the messages are supposed to eventually be passed. The filter will only delay that to ensure a smooth bandwidth usage.

I must have misunderstood the goal with the filter. My assumption was that you would drop packets/requests that are over their quota. In that case, I think the proto is missing some crucial safeguards:

1) Limits on the amount of data that can be buffered. Any unbounded queues are just incidents waiting to happen. 

2) Limits on the queuing delay added by this filter.

I'm not sure how the current iteration of this would handle a scenario with a 100Mb/s limit and a sustained traffic burst in the 1-10Gb/s range. My impression is that latency/memory utilization would get out of hand quickly. You'll almost certainly have to employ some type of active queue management like CoDel or other eviction scheme.

What you're proposing is almost directly analogous to an interface buffer on a network router, so there's decades of work to pull from to avoid common pitfalls. **Probably don't worry about this for now**, but one thing that comes to mind is how you'll schedule the buffered requests to deal with multitenancy issues and misbehaving clients consuming all the bandwidth. I think the current SOA is fq-codel which guarantees fair share usage of the allocated bandwidth among the streams/clients, mitigating the issue to some extent. 

I'm also assuming that by "smoothing" you would like to absorb bursts? One thing that comes to mind is that you don't need to buffer requests, you can just do this via a token bucket that holds more tokens than its refresh amount. The local rate limit filter has an implementation of this. However, in the event of a burst that exceeds the "burst token" amount, there will be drops.

I'll also take a look at some other existing filters and see if it's possible to just extend those to get what you want.

nitgoy commented:
The buffer management will be done via envoy's high and low watermark callbacks as described in this [flow control doc](https://github.com/envoyproxy/envoy/blob/master/source/docs/flow_control.md). The buffer limits are separately configurable at listener and other levels. That way, envoy doesn't need to buffer excessively and will instead disable reading from the source. Check out the StreamRateLimiter code [here](https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/http/fault/fault_filter.cc#L511).

I'll read up more on fq-codel but based on the discussion in the original [issue](https://github.com/envoyproxy/envoy/issues/13604) I was under the impression it's not handled and that's why we still need a bandwidth limiting filter.

nitgoy commented:
Sorry for the long wait but now I have added the filter code and UTs. I'm working on the docs changes and few more tests.

tonya11en commented on source/common/common/token_bucket_impl.cc:
```diff
@@ -40,8 +47,14 @@ std::chrono::milliseconds TokenBucketImpl::nextTokenAvailable() {
   ASSERT(num_tokens <= max_tokens_);
+  absl::WriterMutexLock lock(&mutex_);
+  // Don't reset if reset before and multiple resets aren't allowed.
+  if (reset_once_.has_value() && reset_once_.value()) {
```
I see what you're doing here and why it's necessary for a thread-safe token bucket, but this behavior needs to be explained in comments in the header file. Ideally with an example.

nitgoy replied:
sure. will add explanation.

tonya11en commented on source/common/common/token_bucket_impl.h:
```diff
@@ -29,8 +32,10 @@ class TokenBucketImpl : public TokenBucket {
+  absl::optional<bool> reset_once_;
   MonotonicTime last_fill_;
   TimeSource& time_source_;
+  absl::Mutex mutex_;
```
If you're going to introduce a mutex to guard variables or functions, it would be good to use `ABSL_GUARDED_BY(...)` where relevant.

tonya11en commented on source/extensions/filters/http/common/stream_rate_limiter.h:
```diff
@@ -0,0 +1,93 @@
+#pragma once
```
This refactor should probably be in a separate PR as well.

nitgoy replied:
I was following the change [#13395](https://github.com/envoyproxy/envoy/commit/5d12b675fe13747bc716859275c2d3c21004e99f#diff-1193859f9acae8da94103e4e9210089e429439f5acbd0e6958faeae731a7a9d9) that also did similar refactoring in the same change. I understand the comment for token bucket changes but for this change it should be ok to be reverted if the filter change is reverted.

tonya11en commented on api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto:
```diff
@@ -0,0 +1,62 @@
+
+  // Optional fill_rate for the token buckets (per second). Defaults to 16.
+  // .. note::
+  //   In the current implementation, the fill_rate must be <= 32 to avoid too aggressive refills.
```
Enforce this via the validate.rules

nitgoy replied:
Can you specify multiple validate.rules for the same field? Appreciate any pointers to help with syntax.

tonya11en commented on source/extensions/filters/http/bandwidth_limit/bandwidth_limit.cc:
```diff
@@ -0,0 +1,153 @@
+      fill_rate_(config.has_fill_rate() ? config.fill_rate().value()
+                                        : StreamRateLimiter::DefaultFillRate) {
+  if (per_route && !config.has_limit_kbps()) {
+    throw EnvoyException("bandwidthlimitfilter: limit must be set for per route filter config");
```
I think this should be enforced via the proto validation.

tonya11en commented on source/extensions/filters/http/bandwidth_limit/bandwidth_limit.cc:
```diff
@@ -0,0 +1,153 @@
+    throw EnvoyException("bandwidthlimitfilter: limit must be set for per route filter config");
+  }
+
+  if (fill_rate_ > MaxFillRate) {
```
Same here. Enforce in proto.

tonya11en commented on source/extensions/filters/http/common/stream_rate_limiter.h:
```diff
@@ -0,0 +1,93 @@
+namespace Common {
+
+/**
+ * An HTTP stream rate limiter. Used in the fault filter and bandwidth filter.
```
If it's going to be in the common folder, this should comment be a bit beefier. Why would other filter's use this and what does it do for you, etc.

nitgoy replied:
Sure

tonya11en reviewed (changes requested):
I know you're still working on this, but had a few preliminary comments attached. Main takeaway is that this should be broken into a series of PRs and parts could use some more comments to give it some clarity. I'll admit I didn't look at the filter itself yet, so I'll take a look at that when I make another sweep.

I'm looking forward to the docs though, that'll really help with the big-picture understanding and the **why**s.

nitgoy commented on source/common/common/token_bucket_impl.h:
```diff
@@ -29,8 +32,10 @@ class TokenBucketImpl : public TokenBucket {
+  absl::optional<bool> reset_once_;
   MonotonicTime last_fill_;
   TimeSource& time_source_;
+  absl::Mutex mutex_;
```
I made in-place edits since the new class would be just a copy of this class which itself isn't used in too many places. Apart from fault filter, it's used only by the xDS grpc rate-limit. But I can definitely create a new class if code duplication is not a huge concern.

nitgoy replied:
@tonya11en I have made the token bucket thread-safety optional. Let me know if this approach looks good to you or not and I'll create a separate PR for these changes.

nitgoy replied:
Created new PR for the token bucket changes: https://github.com/envoyproxy/envoy/pull/14827

nitgoy commented on source/common/common/token_bucket_impl.h:
```diff
@@ -5,10 +5,12 @@
 
 /**
- * A class that implements token bucket interface (not thread-safe).
+ * A class that implements token bucket interface.
```
Sure. I'll move the token bucket changes to a separate PR.

nitgoy commented:
As suggested, I've created two separate PRs:
Token bucket: https://github.com/envoyproxy/envoy/pull/14827
Stream rate limiter: https://github.com/envoyproxy/envoy/pull/14828

mattklein123 commented:
Sorry I haven't looked at this yet, but will put this in waiting until the base PRs are merged and then we can circle back.

/wait

leoribeiro2 commented:
Hi guys!
Have you any updates about that feat?

nitgoy commented:
> Hi guys!
> Have you any updates about that feat?
> 
> @nitgoy

Sorry it's taken so long. We're now using it for our product so that helped to uncover any functional issues. Hoping to push out new changes pretty soon. Most likely after completing this dependent PR: https://github.com/envoyproxy/envoy/pull/14828

phlax commented on api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto:
```diff
@@ -0,0 +1,62 @@
+
+import "google/protobuf/wrappers.proto";
+
+import "udpa/annotations/migrate.proto";
```
this isnt currently used

nitgoy replied:
do you mean this is deprecated or it's not being used in this particular PR? Guessing you meant the former since I can't find the doc which used to have a mention about using this with '*alpha' protos.

phlax replied:
this proto file doesnt use it so it emits a warning in CI

the current docs have just been updated to remove reference to `migrate.proto` here - https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md#adding-an-extension-configuration-to-the-api

nitgoy replied:
got it. removed these imports.

phlax commented on api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto:
```diff
@@ -0,0 +1,62 @@
+
+import "udpa/annotations/migrate.proto";
+import "udpa/annotations/status.proto";
+import "udpa/annotations/versioning.proto";
```
this one also isnt used

nitgoy replied:
removed

nitgoy commented:
I have opened a replacement PR for this one after addressing most of the comments: https://github.com/envoyproxy/envoy/pull/16358. I was running into some DCO issues which were getting non-trivial to fix. Please take a look at the new PR, will close this one.
